numpy/fft/_pocketfft_umath.c

```diff
@@ -233,17 +238,24 @@ irfft_loop(char **args, npy_intp const *dimensions, npy_intp const *steps, void
         double fct = *(double *)fp;
         char *op_or_buff = buff == NULL ? op : buff;
         double *op_double = (double *)op_or_buff;
-        /* Copy complex input to buffer in FFTpack order */
-        op_double[0] = ((double *)ip)[0];
-        if (nin > 1) {
-            copy_data(ip + step_in, step_in, nin - 2,
-                      (char *)&op_double[1], sizeof(npy_cdouble), npts / 2 - 1,
+        /*
+         * Copy complex input to buffer in FFTpack order:
+         * R0,R1,I1,...Rn-1,In-1,Rn[,In] (last for npts odd)
```

Comment:
I got tripped up reading this that there's no `I0`, maybe a comment explicitly saying that the first entry is defined to be real?

Reply:
The data is often packed a bit further in the even case, `R0, RN, R1, I1 ...` .  Don't know what fftpack did, but the form NumPy uses differed from the form SciPy returned back in the day. Having all complex numbers in the proper order and alignment is much less confusing.

Reply:
Yes, the order @charris mentions is that used in Numerical Recipes, which I agree is nicer. But that's not what pocketfft and fftpack do...

Anyway, yes, happy to add a comment!  Since this code is for an irfft, by the assumption that the input values to the forward fft were real, the zero frequency element, which is just the sum of all the input elements, must be real too. For even number of points, by symmetry, the Nyquist frequency is also real (and of course just by counting the number of degrees of freedom one reaches the same conclusion).